email-encryption/src/com/axonivy/market/emailencryption/service/EncryptedEmailSender.java

```diff
@@ -0,0 +1,207 @@
+		bodyMultipart.addBodyPart(bodyMessagePart);
+
+		for (File attachment: mail.getAttachments()) {
+			try {
+				addAttachment(bodyMultipart, attachment);
+			} catch (UnsupportedEncodingException e) {
+				Ivy.log().error(e.toString());
+			}
+		}
+		// Get a Session object and create the mail message
+		Properties props = null;
+		Session session = null;
+
+		try {
+			MailClientConfig emailConfig = mailSetup();
+			props = getProperties(emailConfig);
+			session = getSession(props, emailConfig);
+		} catch (Exception e) {
+			throw new IvyException(e);
+		}
+
+		if (Security.getProvider(BouncyCastleProvider.PROVIDER_NAME) == null) {
+			Security.insertProviderAt(new BouncyCastleProvider(), 1);
+		}
+
+		SMIMEEnvelopedGenerator gen = new SMIMEEnvelopedGenerator();
+		JceKeyTransRecipientInfoGenerator keyGen = new JceKeyTransRecipientInfoGenerator(certificate);
+		keyGen.setProvider(BouncyCastleProvider.PROVIDER_NAME);
+		gen.addRecipientInfoGenerator(keyGen);
+		OutputEncryptor encryptor = new JceCMSContentEncryptorBuilder(CMSAlgorithm.RC2_CBC)
+				.setProvider(BouncyCastleProvider.PROVIDER_NAME).build();
+
+		MimeMessage messageToEncrypt = new MimeMessage(session);
+		messageToEncrypt.setContent(bodyMultipart);
+		messageToEncrypt.saveChanges();
+		MimeBodyPart contentPart = gen.generate(messageToEncrypt, encryptor);
+
+		Address fromUser = new InternetAddress(StringUtils.remove(mail.getFrom(), ";"));
+		Address toUser = new InternetAddress(StringUtils.remove(mail.getTo(), ";"));
+
+		MimeMessage message = new MimeMessage(session);
+		message.setFrom(fromUser);
+		message.setRecipient(Message.RecipientType.TO, toUser);
+		message.setSubject(mail.getSubject());
+		message.setContent(contentPart.getContent(), contentPart.getContentType());
+		message.saveChanges();
+
+		Transport.send(message);
+		Thread.currentThread().setContextClassLoader(csBackup);
+	}
+
+	/**
+	 * Adds attachment to multipart
+	 * 
+	 * @param multipart
+	 * @param ivyFile
+	 * @throws MessagingException
+	 * @throws UnsupportedEncodingException
+	 * @throws SMIMEException
+	 */
+	private static void addAttachment(MimeMultipart multipart, File ivyFile)
+			throws MessagingException, UnsupportedEncodingException, SMIMEException {
+		DataSource source = new FileDataSource(ivyFile.getJavaFile());
+		MimeBodyPart messageBodyPart = new MimeBodyPart();
+		messageBodyPart.setDataHandler(new DataHandler(source));
+		messageBodyPart.setFileName(ivyFile.getName());
+
+		multipart.addBodyPart(messageBodyPart);
+	}
+
+	/**
+	 * Gets session for mail sending.
+	 * 
+	 * @param props
+	 * @param emailConfig
+	 * @throws Exception
+	 */
+	private static Session getSession(Properties props, MailClientConfig emailConfig) throws Exception {
+		if (emailConfig.encryption() == EmailEncryption.SSL || emailConfig.encryption() == EmailEncryption.START_TLS) {
+			Authenticator authenticator = new Authenticator() {
+				public PasswordAuthentication getPasswordAuthentication() {

```

Comment:
as far as I know this is not good. the authentication should be made when connecting with transport:
transport.connect(user, password);

As far as I know the authenticator was made for rich ui's

Reply:
Hi @alexsuter ,
thank you for your inputs. I have an other implementation with JavaMailSender from Spring. If you think, it can be an option, I can adapt the code and use the JavaMailSender instead.

Reply:
You can try to use Sprint, but please note that in most cases you will also require a big dependency tree of third party libraries. And some of them may be already in the ivy-core ... so you'll have less homerbrew code here, but more work to keep the dependencies compliant.
At least it was so in our last attempts to use spring-libraries. If we are lucky the JavaMailSender is a good exceptoin.

Reply:
Hi @ivy-rew ,
the implementation with JavaMailSender is now done. can you please review it again?